### ISSUE TITLE
Example/docs for string_list_setting

### ISSUE THREAD

**Issue by Michaelhobo** [bazelbuild/bazel-skylib#296]
Is there more documentation (preferably with examples) for using string_list_setting? I'm trying to pass defines to a cc_library rule, like this:

```
cc_library(
  name = "...",
  defines = ":nrf_defines",
)

string_list_setting(
  name = "nrf_defines",
  build_setting_default = [],
)
```

When I build this, cc_library rules spit out the error:

```
expected value of type 'list(string)' for attribute 'defines' in 'cc_library' rule, but got ":nrf_defines" (string)
```

This also makes sense, since the types are different, but I don't know how to proceed. How do I convert the string_list_setting into a list of strings when filling in the `defines` field of a cc_library?

**Comment by tetromino** [comment]
Fair point - we don't have much documentation for `string_list_setting`, or really for any part of `common_settings`.

Pinging @gregestren from Bazel Configurability for advice.

**Comment by gregestren** [comment]
https://github.com/bazelbuild/examples/tree/master/rules/starlark_configurations is probably the best spot for examples, linked from the docs at https://docs.bazel.build/versions/master/skylark/config.html#predefined-settings.

In your case, the biggest challenge is that Starlark settings (which `string_list_setting` is) work most thoroughly with Starlark rules. `cc_library` is a builtin rule, so it doesn't know how to unpack them.

There are some workarounds, but it depends on exactly what you want to do. how many values do you imagine setting in the `defines`? Anything arbitrary? Just a pre-known set of possibilities?

**Comment by Michaelhobo** [comment]
> There are some workarounds, but it depends on exactly what you want to do. how many values do you imagine setting in the `defines`? Anything arbitrary? Just a pre-known set of possibilities?

There are a number of different options (10+ different defines) that are set using -D in the nrf52 SDK, so there are technically a lot of possibilities. These include enabling timers and other features, stack/heap size, board model, as well as user-defined settings. Once set, I don't think they change very often though, so if we have to boil the possibilities down into a set of common ones, that's a possibility, as long as the user can change them. 

What workarounds are you thinking of?

**Comment by gregestren** [comment]
The main workaround I was thinking of is simply:

```
cc_binary / cc_library
    defines = select({...})
```

which works when the total space of possible values is known and small.

But I don't think that works for your case, nor with wanting to define the values directly next to the `cc_binary`.

What if you have the transition at your `cc_binary` set [`--copt`](https://docs.bazel.build/versions/4.0.0/user-manual.html#flag--copt)? You should be able to add `-Dkey=val` values for `--copt`, and those should be consumed by `cc_library`s below just as if they had directly set `defines`.

You could even trigger the transition on `nrf_defines` if you still wanted to control the whole thing from `nrf_defines` specifically.

**Comment by Michaelhobo** [comment]
> What if you have the transition at your `cc_binary` set [`--copt`](https://docs.bazel.build/versions/4.0.0/user-manual.html#flag--copt)? You should be able to add `-Dkey=val` values for `--copt`, and those should be consumed by `cc_library`s below just as if they had directly set `defines`.

Ah, that should solve my problem perfectly! Thank you, I guess I should've read the documentation for copts a bit better. I'll try it soon, but I'm pretty sure that'll solve my problem.

Even if my immediate problem is solved, it seems like we still want better documentation - should we leave this issue open for that?

**Comment by tetromino** [comment]
Yes - a PR improving the docs is still very much welcome

**Comment by gregestren** [comment]
Agreed.

**Comment by Michaelhobo** [comment]
Actually, according to this doc, copts doesn't propagate to its dependencies: https://docs.bazel.build/versions/master/be/c-cpp.html#cc_binary

> Each string in this attribute is added in the given order to COPTS before compiling the binary target. The flags take effect only for compiling this target, not its dependencies, so be careful about header files included elsewhere. All paths should be relative to the workspace, not to the current package.

I believe that means, if I have this:

```
nrf_cc_binary(
  copts = [
    "-DA=B",
  ],
)
```

This will only work for compiling that target, but all the cc_library deps won't have this when it's defined.

**Comment by Michaelhobo** [comment]
Ok, after reading bazel docs some more, I see there's a --copt tool flag, and I should be able to set it with `//command_line_option:copt`.

**Comment by tkanarsky** [comment]
sorry for necro -- this came up in a google search, and https://github.com/bazelbuild/examples/tree/master/rules/starlark_configurations is 404ing. is there an updated link?

**Comment by gregestren** [comment]
No worries. Relevant examples should be at https://github.com/bazelbuild/examples/tree/main/configurations now.
